### PR TITLE
qa: exercise metadata replication in workloads

### PR DESCRIPTION
**Pull request by batrick**
Fixes: https://tracker.ceph.com/issues/51332
Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by batrick** [comment]
@liewegas this should not be too crazy a configuration, right? i.e. trying to get every directory to replicate.

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-July-04

**Comment by batrick** [comment]
Didn't see the replication I was expecting in

/ceph/teuthology-archive/pdonnell-2021-07-04_02:32:34-fs-wip-pdonnell-testing-20210703.052904-distro-basic-smithi/6251275

and

/ceph/teuthology-archive/pdonnell-2021-07-04_02:32:34-fs-wip-pdonnell-testing-20210703.052904-distro-basic-smithi/6251275/remote/smithi198/log/ceph-mds.l.log.gz

Will debug this further later...

File: qa/suites/fs/workload/replication/default.yaml [hunk 1]
```diff
@@ -0,0 +1 @@
+# Use default (8000)
```
Comment (vshankar) [inline qa/suites/fs/workload/replication/default.yaml:1]: default.yaml is not _really_ needed, isn't it?

... or was it the reason metadata was not replicated as expected? (that would be weird).

Comment (vshankar) [reply qa/suites/fs/workload/replication/default.yaml:1]: ah ok.

I assumed that we _always_ want to exercise metadata replication for tests.

**Comment by batrick** [comment]
@vshankar see latest commits. Basically replication almost never happens with the current code. I'm going to do some testing with these changes to see if that has the expected effect of stimulating more replication.

Testing is good!

**Comment by batrick** [comment]
New test run in 3h: https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20220218.000925

**Comment by batrick** [comment]
Oh, and here is the previous run: https://pulpito.ceph.com/pdonnell-2022-02-17_23:13:18-fs:workload-wip-pdonnell-testing-20220217.200926-distro-default-smithi/

````
2022-02-18T00:09:24.271+0000 7fcfc3b5c700 20 mds.0.bal hit_dir 2 pop is 17, frag 000* size 17 [pop IRD:[C 0.00e+00] IWR:[C 0.00e+00] RDR:[C 1.70e+01] FET:[C 0.00e+00] STR:[C 0.00e+00] *LOAD:17.0]

2022-02-18T00:09:24.271+0000 7fcfc3b5c700 20 mds.0.bal hit_dir 2 pop 170 spread in [dir 0x10000005ff9.000* /client.0/tmp/t/linux-5.4/block/ [2,head] auth{2=1} v=723 cv=677/677 REP state=1610874881|complete f(v30 m2022-02-18T00:00:39.860659+0000 17=17+0) n(v10 rc2022-02-18T00:00:39.860659+0000 b371593 17=17+0) hs=17+2,ss=0+0 dirty=11 | ptrwaiter=0 child=1 frozen=0 subtree=0 importing=0 replicated=1 dirty=1 authpin=0 0x564e820e6000]
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  7 mds.0.cache sending dir_update on [dir 0x10000005ff9.000* /client.0/tmp/t/linux-5.4/block/ [2,head] auth{2=1} v=723 cv=677/677 REP state=1610874881|complete f(v30 m2022-02-18T00:00:39.860659+0000 17=17+0) n(v10 rc2022-02-18T00:00:39.860659+0000 b371593 17=17+0) hs=17+2,ss=0+0 dirty=11 | ptrwaiter=0 child=1 frozen=0 subtree=0 importing=0 replicated=1 dirty=1 authpin=0 0x564e820e6000] bcast 1 to 0,1,3,4
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  7 mds.0.cache sending dir_update on [dir 0x10000005ff9.000* /client.0/tmp/t/linux-5.4/block/ [2,head] auth{2=1} v=723 cv=677/677 REP state=1610874881|complete f(v30 m2022-02-18T00:00:39.860659+0000 17=17+0) n(v10 rc2022-02-18T00:00:39.860659+0000 b371593 17=17+0) hs=17+2,ss=0+0 dirty=11 | ptrwaiter=0 child=1 frozen=0 subtree=0 importing=0 replicated=1 dirty=1 authpin=0 0x564e820e6000] to 1
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  1 -- [v2:172.21.15.187:6838/2062969492,v1:172.21.15.187:6839/2062969492] send_to--> mds [v2:172.21.15.187:6836/3389520465,v1:172.21.15.187:6837/3389520465] -- dir_update(0x10000005ff9.000*) v1 -- ?+0 0x564e818b4000
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  1 -- [v2:172.21.15.187:6838/2062969492,v1:172.21.15.187:6839/2062969492] --> [v2:172.21.15.187:6836/3389520465,v1:172.21.15.187:6837/3389520465] -- dir_update(0x10000005ff9.000*) v1 -- 0x564e818b4000 con 0x564e7d685400
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  7 mds.0.cache sending dir_update on [dir 0x10000005ff9.000* /client.0/tmp/t/linux-5.4/block/ [2,head] auth{2=1} v=723 cv=677/677 REP state=1610874881|complete f(v30 m2022-02-18T00:00:39.860659+0000 17=17+0) n(v10 rc2022-02-18T00:00:39.860659+0000 b371593 17=17+0) hs=17+2,ss=0+0 dirty=11 | ptrwaiter=0 child=1 frozen=0 subtree=0 importing=0 replicated=1 dirty=1 authpin=0 0x564e820e6000] to 3
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  1 -- [v2:172.21.15.187:6838/2062969492,v1:172.21.15.187:6839/2062969492] send_to--> mds [v2:172.21.15.161:6838/3448486655,v1:172.21.15.161:6839/3448486655] -- dir_update(0x10000005ff9.000*) v1 -- ?+0 0x564e85023e00
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  1 -- [v2:172.21.15.187:6838/2062969492,v1:172.21.15.187:6839/2062969492] --> [v2:172.21.15.161:6838/3448486655,v1:172.21.15.161:6839/3448486655] -- dir_update(0x10000005ff9.000*) v1 -- 0x564e85023e00 con 0x564e7d685c00
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  7 mds.0.cache sending dir_update on [dir 0x10000005ff9.000* /client.0/tmp/t/linux-5.4/block/ [2,head] auth{2=1} v=723 cv=677/677 REP state=1610874881|complete f(v30 m2022-02-18T00:00:39.860659+0000 17=17+0) n(v10 rc2022-02-18T00:00:39.860659+0000 b371593 17=17+0) hs=17+2,ss=0+0 dirty=11 | ptrwaiter=0 child=1 frozen=0 subtree=0 importing=0 replicated=1 dirty=1 authpin=0 0x564e820e6000] to 4
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  1 -- [v2:172.21.15.187:6838/2062969492,v1:172.21.15.187:6839/2062969492] send_to--> mds [v2:172.21.15.187:6832/2989586981,v1:172.21.15.187:6833/2989586981] -- dir_update(0x10000005ff9.000*) v1 -- ?+0 0x564e85023c20
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  1 -- [v2:172.21.15.187:6838/2062969492,v1:172.21.15.187:6839/2062969492] --> [v2:172.21.15.187:6832/2989586981,v1:172.21.15.187:6833/2989586981] -- dir_update(0x10000005ff9.000*) v1 -- 0x564e85023c20 con 0x564e7d65dc00
2022-02-18T00:09:24.271+0000 7fcfc3b5c700  7 mds.0.server reply_client_request 0 ((0) Success) client_request(client.25187:209963 readdir #0x10000005ff9 2022-02-18T00:09:24.270133+0000 caller_uid=1000, caller_gid=1238{6,36,1000,1238,}) v4
````

From: /ceph/teuthology-archive/pdonnell-2022-02-17_23:13:18-fs:workload-wip-pdonnell-testing-20220217.200926-distro-default-smithi/6692672/remote/smithi187/log/866b1250-904b-11ec-8c35-001a4aab830c/ceph-mds.l.log.gz

Fix looks good at the moment.

**Comment by batrick** [comment]
TODO: reevaluate default for mds_bal_replicate_threshold.

**Comment by vshankar** [comment]
@mchangir This might be interesting for testing the un-inlining changes for tests...

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
going to reopen this, the PR is old and some jenkins pr tests hate it

**Pull request closed without merging**